tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/EtcExpr.scala

```diff
@@ -0,0 +1,117 @@
+package at.forsyte.apalache.tla.typecheck.etc
+
+import at.forsyte.apalache.tla.typecheck.TlaType1
+
+/**
+  * An expression in a simple typed lambda calculus. Here we do not care about the concrete values,
+  * but only care about the types that the expressions carry. Since we introduce type variables and quantifiers
+  * in the principal types, this calculus is probably not that simple. If you have an idea, whether it is in
+  * System F, or System F_\omega, or whatever, let us know.
+  *
+  * @author Igor Konnov
+  */
+sealed trait EtcExpr {
+  /**
+    * The reference to the source.
+    * This identifier is not taken into account in equals and hash.
+    */
+  val sourceRef: EtcRef
+}
+
+/**
+  * A constant expression, i.e., just a type (may be polymorphic).
+  *
+  * @param polytype a type constant that may have free variables (polytype).
+  * @param sourceRef the identifier of the TLA+ expression that resulted in this STCExpr (ignored in equals).
+  */
+case class EtcConst(polytype: TlaType1)(val sourceRef: EtcRef) extends EtcExpr {
+  override def toString: String = {
+    s"$sourceRef@$polytype"
+  }
+}
+
+/**
+  * A reference to a name, which can be introduced in the initial type context, or with STCAbs.
+  * Note that name is not a type variable, but rather a TLA+ name that carries a type.
+  *
+  * @param name  a name
+  * @param sourceRef the identifier of the TLA+ expression that resulted in this STCExpr (ignored in equals).
+  */
+case class EtcName(name: String)(val sourceRef: EtcRef) extends EtcExpr {
+  override def toString: String = s"$sourceRef@$name"
+}
+
+/**
+  * Lambda abstraction that introduces an operator type
+  * of multiple arguments that range over their respective types Set(*).
+  *
+  * @param body          the function body
+  * @param paramsAndDoms parameter names and type expressions that encode sets of values
+  * @param sourceRef     the identifier of the TLA+ expression that resulted in this STCExpr (ignored in equals).
+  */
+case class EtcAbs(body: EtcExpr, paramsAndDoms: (String, EtcExpr)*)(val sourceRef: EtcRef) extends EtcExpr {
```

Comment:
Why does this take body before the arguments? It seems counter intuitive to me given the usual syntax for lambdas, which always take the argument first.

Reply:
Right. That is our standard hack to use variadic operators in Scala. You can construct `EtcAbs` with at least one argument.

Reply:
Oh. I see. Why not just have `EtcAbs` take a sequence/list for the args?